docs/midstream-getting-serveless-operator-bugfixes.md

```diff
@@ -0,0 +1,132 @@
+- If using Knative Serving, permission to delete pods and replicasets in `knative-serving`, `knative-serving-ingress` namespaces.
+
+## 3. Execute/Resolution
+
+1. Get the Serverless Operator images used in the OpenShift cluster right now, for later comparison:
+  ```shell
+  > kubectl get pods -n openshift-serverless -o jsonpath='{range .items[*]}{"\n"}{.metadata.name}{": "}{range .status.containerStatuses[*]}{.imageID}{end}{end}' | sort 
+  knative-openshift-5dbb4bd4b7-q6wdp: registry.ci.openshift.org/knative/openshift-serverless-v1.24.0@sha256:26c3127e3dab1e102b94c305916d405b57d397c56348e0d73c7e1c2cddf3310b
+  knative-openshift-ingress-78766bdc5c-zfzfl: registry.ci.openshift.org/knative/openshift-serverless-v1.24.0@sha256:7a417307328f76462560c5d4f814566135e59fc1b20758be720d090047ec682e
+  knative-operator-65487bf7fc-vbgkk: registry.ci.openshift.org/knative/openshift-serverless-v1.24.0@sha256:589735abf033c61247e594f3fba9e7d87ac1093d1ebc0abe73f825944fe5e465
+  ```
+2. Delete Serverless Operator pods to force them to be recreated with the new image (we explicitly set `imagePullPolicy` to `Always`):
+  ```shell
+  > kubectl delete pods -n openshift-serverless -l 'name in (knative-operator, knative-openshift, knative-openshift-ingress)'
+  pod "knative-openshift-555d4d98d7-8ns5b" deleted
+  pod "knative-openshift-ingress-7bc7c7b47b-gc5c9" deleted
+  pod "knative-operator-67c4958cc6-zg8b9" deleted
+  ```
+
+3. If using Knative Serving, delete Knative Serving control plane replicasets. They will be recreated by the operator:
+  ```shell
+  > kubectl delete replicasets.apps -n knative-serving -l 'app in (controller, webhook)'
+  replicaset.apps "controller-57466669cf" deleted
+  replicaset.apps "controller-7bb7748cd8" deleted
+  replicaset.apps "webhook-6cb8b848bd" deleted
+  ```
+
+4. If using Knative Serving, delete jobs as their images cannot be mutated:
+  ```shell
+  > kubectl delete jobs -n knative-serving -l 'app in (storage-version-migration-serving)'
+  job.batch "storage-version-migration-serving-serving-1.3.0" deleted
+  ```
+
+5. If using Knative Serving, delete Knative Serving Ingress control plane replicasets. They will be recreated by the operator:
+  ```shell
+  > kubectl delete replicasets.apps -n knative-serving-ingress -l 'app in (net-kourier-controller)'
+  replicaset.apps "net-kourier-controller-84d4b75589" deleted
+  ```
+
+6. If using Knative Eventing and Knative Kafka components, delete Knative Eventing and Knative Kafka control plane replicasets. They will be recreated by the operator:
+  ```shell
+  > kubectl delete replicasets.apps -n knative-eventing -l 'app in (kafka-controller, eventing-controller, eventing-webhook, kafka-webhook-eventing)'
+  replicaset.apps "eventing-controller-5999f874f8" deleted
+  replicaset.apps "eventing-webhook-86dd7d855b" deleted
+  replicaset.apps "kafka-controller-6bd78d9f4f" deleted
+  replicaset.apps "kafka-webhook-eventing-8444b7ccb4" deleted
+  ```
```

Comment:
what about the data plane?

Reply:
The upstream workloads use `ImagePullPolicy=IfNotPresent`, so I don't think that the new images are gonna get used if the new pods are scheduled onto the same nodes as the previous ones.

Reply:
Yeah, I noted in the PR description that this is about getting fixes in SO operator itself and not the Knative control plane / data plane.

I added instructions to delete replicasets for Knative control plane because I've seen there were some "stuck" replicasets even when there were no image changes for them. It may be an edge case, but I guess it won't hurt.

More specifically, the constant loop for reconciling was happening because there were some replicasets with difference.

Let's create the SOP for getting Knative image fixes in a separate place (or we can add this to this SOP, but later). We need a "force" mechanism to pull the images for them anyway.

Reply:
> More specifically, the constant loop for reconciling was happening because there were some replicasets with difference.

Was it still happening even after the Matthias fix was backported?
It seems worth investigating what exactly was going on because it might be hiding a bug.

Reply:
I couldn't figure out why it was happening.
- The operator image was updated
- Deployment was set properly (no fluctuating `--disable-controller` arg)
- Operator was not touching the deployment after it touched it once

However, the old replicasets were still there.

Not sure if I can actually reproduce this.

Also not sure if this is a bug in the operator. It seemed more like a Kubernetes bug to me. ...or, how deployments<>replicasets work.

Reply:
What I mean specifically is, I see this in operator deployment: `revisionHistoryLimit: 1`

However, there were at least 3 replicasets.